devnotes/tpm.md

```diff
@@ -0,0 +1,228 @@
+  * Open implementation.
+
+* **Cons**
+  * Need to secure the BMC and manage that properly.
+
+* **Risks**
+  * BMC might not have been designed to fulfill security requirements.
+
+* **Implementation effort**
+  * Implement fTPM in BMC.
+  * LPC gadget BMC -> POWER.
+  * No existing driver. Maybe need to hack up existing LPC driver.
+
+* **Other notes**
+  * We could go full RO with coreboot if desired (by sacrificing some
+    performance for MVPD).
+
+******
+### FPGA translator (LPC TPM cycle - LPC TPM cycle)
+
+New piece of hardware that addresses LPC cycles issues.
+
+* **Pros**
+  * Conventional TPM modules.
+  * No need for a new driver.
+
+* **Cons**
+  * Need to design and manufacture this hardware.
+
+* **Risks**
+
+* **Implementation effort**
+  * Slightly easier than the fTPM on BMC - but not much.
+
+* **Other notes**
+
+******
+### FPGA translator (SPI - LPC TPM cycle)
+
+New piece of hardware that addresses LPC cycles issues.
+
+* **Pros**
+  * Conventional TPM modules.
+  * No need for a new driver.
+
+* **Cons**
+  * Need to design and manufacture this hardware.
+
+* **Risks**
+
+* **Implementation effort**
+  * Slightly easier than the fTPM on BMC - but not much.
+
+* **Other notes**
+
+******
+### Open TPM on MCU (e.g. [lpnTPM])
+
+Find an existing open hardware TPM and possibly extend it for I2C interface.
+
+* **Pros**
+  * Open hardware.
+
+* **Cons**
+  * Need to implement and manufacture TPM module.
+
+* **Risks**
+  * I2C is not in the scope as of now.
+
+* **Implementation effort**
+  * Slightly easier than the fTPM on BMC - but not much.
+
+* **Other notes**
+
+******
+### TPM on [LibreBMC][LibreBMC-site]
+
+Embed TPM into an alternative BMC hardware that's still in development?
+The BMC itself is implemented as a custom extension card defined by DS-CMS?
+
+* **Pros**
+  * Open hardware.
+
+* **Cons**
+  * Project is in an early phase of development.
+
+* **Risks**
+  * Current design does not have TPM connector.
+
+* **Implementation effort**
+
+* **Other notes**
+  * Maybe we can implement fTPM in FPGA BMC.
+
+******
+### fTPM in [SBE][SBE-site]
+
+Implement TPM on a chip that's used primarily for booting POWER9 processor.
+
+* **Pros**
+  * SBE chip is an independent piece of hardware that's part of secure boot
+    sequence for POWER9.
+
+* **Cons**
+
+* **Risks**
+  * SBE SEEPROM may not have enough free space to include fTPM implementation.
+
+* **Implementation effort**
+  * Will need a driver for the OS.
+    If there is one for SBE communication from IBM, it still needs an update to
+    support custom SBE firmware and extensions made to its protocol.
+
+* **Other notes**
+  * We have not dealt with SBE much so far, but coreboot uses it less than
+    Hostboot does, so we might be able to remove unused parts of sources.
+  * Looks like `sbe_seeprom_DD2` is compressed from 86 to 65 KiB, but compressed
+    `sbe_seeprom_AXONE` is 150 KiB.
+  * There are 96 KiB of RAM and 256 KiB of SEEPROM.
```

Comment:
SEEPROM has ECC and is split into 64KB chunks in a way that no ECC word crosses that boundary, which makes it < 228KB. 20KB of that is reserved for HBBL, and IIRC there is also some rings data.

Reply:
According to SEEPROM dump I did some time ago, there is ~200 bytes left with ECC, ~175 without it.